Implemented rudimentary inbreeding-prevention logic

kausha1n opened the pull request:
- Refactored initial population creation logic to account for parents and children
- Moved assigning of gender to Male or Female to respective constructor methods
- Implemented logic to check for in-breeding. Current logic checks for half/full siblings and direct descendants. Have to implement logic for 'Nth cousin kth removed'(see https://www.metaspoon.com/wp-content/uploads/2016/12/familychart.gif ). Caveats: due to increased restriction to breed, male population tends to end up in single digit number often times. Further optimization needed

kausha1n committed:
Implemented rudimentary inbreeding-prevention logic

- Refactored initial population creation logic to account for parents and children
- Moved assigning of gender to Male or Female to respective constructor methods
- Implemented logic to check for in-breeding. Current logic checks for half/full siblings and direct descendants. Have to implement logic for 'Nth cousin kth removed'(see https://www.metaspoon.com/wp-content/uploads/2016/12/familychart.gif ). Caveats: due to increased restriction to breed, male population tends to end up in single digit number often times. Further optimization needed

manishlamichhane commented:
It actually is good that male converges to 1 I think. He is the fittest of all. 
One of the use of genetic-algo is to maximize attribute given other attribute.
Here we maximize age, provided other attributes.

9radiopassive0 commented on nature.py:
```diff
@@ -1,13 +1,52 @@
-        self.female_population = [Female(gender=Gender.FEMALE) for i in range(50)]
-    
+        # Define number of original ancestors
+        orig_male_ancestors_pop = int((Ancestors.MALE_PCT*Ancestors.ORIG_POP)/100)
```
Percentage calculation is wrong. Should be:

int((Ancestors.MALE_PCT/Ancestors.ORIG_POP *100))

kausha1n replied:
It is correct: (x percentage of total) = (x * total)/100

9radiopassive0 commented on individual.py:
```diff
@@ -48,12 +59,12 @@ def __init__(self, **kwargs):
         self.is_pregnent = False
         self.feminity = random.choice(range(0, 100, 5))  # NOTE(Manish): determins sex of off-spring
-        super().__init__(**kwargs)
+        super().__init__(gender=Gender.FEMALE, **kwargs)
```
I think Gender attribute should be part of Individual. Both Female n Male can inherit it. Else we repeat the same code in two places no? Any advantage of having them in child level constructor?

kausha1n replied:
A Male object will always have gender Gender.MALE, and a Female object will always have gender Gender.FEMALE. That information is universally known, we should not need to further tell the program that a Male is of MALE gender type, unless it is intended for other purpose as well(like LGBTQ). Even if we account for LGBTQ, we will only end up having 2 sexes relevant to a procreation process in our use-case. Having to tell each time that a Male object is of MALE gender type is redundant, and discards the core essence of inheritance. Gender attribute is still a part of Individual, we have the power to pre-assign it to the correct type since we are using inheritance

kausha1n committed:
[fix] inbreeding-prevention

- the logic checks for family member to permit/disallow procreation based on distance between two parties(DFS)
